v2/devdoc/logging.md

```diff
@@ -0,0 +1,162 @@
+# `c-logging` requirements
+
+## Basic functionality
+
+`c-logging` shall expose APIs allowing the user to log events that can be used for observing the software activity, debugging, etc.
+
+The exposeed API shall allow creating contexts both dynamically and on the stack and using those contexts for logging.
+
+The API shall allow easy formatting of values to be placed in the contexts by using printf-like formatting.
+
+```
+void logger_log(LOG_LEVEL log_level, LOG_CONTEXT_HANDLE log_context, const char* format, ...);
+
+// macro that can be used to create a dynamically allocated context
+#define LOG_CONTEXT_CREATE(dest_log_context, parent_context, ...) \
+    ...
+
+// destroy a dynamically allocated context
+#define LOG_CONTEXT_DESTROY(log_context_handle) \
+    ...
+
+// macro that can be used to create a context on the stack
+#define LOCAL_LOG_CONTEXT_DEFINE(log_context, parent_context, ...) \
+    ...
+
+// when a user wants to specify which fields should be formatted in a context, they can use LOG_CONTEXT_FIELD
+// LOG_CONTEXT_FIELD(field_name, field_conversion_specifier, field_value)
+```
+
+Example usage:
+
+```c
+    // chained allocated context
+    LOG_CONTEXT_HANDLE context_1;
+    LOG_CONTEXT_CREATE(context_1, NULL, LOG_CONTEXT_FIELD("name", "%s", "haga"));
+    LOG_CONTEXT_HANDLE context_2;
+    LOG_CONTEXT_CREATE(context_2, context_1, LOG_CONTEXT_FIELD("last name", "%s", "uaga"), LOG_CONTEXT_FIELD("age", "%d", 42));
+
+    logger_log(LOG_LEVEL_VERBOSE, context_1, "log context 1 data");
+    logger_log(LOG_LEVEL_VERBOSE, context_2, "log context 2 data");
+
+    LOG_CONTEXT_DESTROY(context_1);
+
+    // stack allocated context
+    LOCAL_LOG_CONTEXT_DEFINE(local_context, context_2, LOG_CONTEXT_FIELD("the knights that say", "%s", "Nee!"));
+
+    logger_log(LOG_LEVEL_VERBOSE, &local_context, "log with local context");
+
+    LOG_CONTEXT_DESTROY(context_2);
+```
+
+## Log levels
+
+The following log levels shall be exposed by the library:
+
+- `CRITICAL`
+
+- `WARNING`
+
+- `ERROR`
+
+- `INFO`
+
+- `VERBOSE`
+
+## Sinks
+
+Multiple sinks shall be supported. Setting up globally which sinks to be used shall be supported (so that they do not have to be specified every time a `c_logging_log` call is made).
+
+Defining the sinks to use shall be done by defining a NULL terminated array, where each entry shall be a pointer to a structure that represents the sink interface.
```

Comment:
>by defining a NULL terminated array [](http://example.com/codeflow?start=40&length=35)

Interesting, in which context is sizeof(log_sinks) not available... 

Will this be passed on the stack?

Reply:
I was thinking at some point of allowing the configuration at link time.
As in having the user have a variable that is an array with the sinks. An the linker makes the connection - that's why the NULL terminated array.

For now it is anyway just an array of these sinks in a .c file in the logging lib. And I think we should start with such simple thing and then add if we need more.